### PR TITLE
fix(ivy): support queries for views inserted in lifecycle hooks

### PR DESCRIPTION
**Pull request by pkozlowski-opensource**
This PR changes how we generate code for view queries. 

Previously query-related instructions (create, refresh) were part of component's template function. Unfortunately this meant that queries are refreshed too early (before change detection hooks are run) and potentially were missing views inserted by directives - more details in #23707

With this PR query-related instructions are generated in a dedicated `queryInstructions` function (similar to a template function). The `queryInstructions` function has the same structure as the main template function but is executed at different time.

### CODE REVIEW

File: packages/core/src/render3/STATUS.md [hunk 1]
```diff
@@ -192,7 +192,7 @@ The goal is for the `@Component` (and friends) to be the compiler of template. S
 | `@ContentChildren`              |  ✅     |  ✅      |  ❌       |
 | `@ContentChild`                 |  ✅     |  ✅      |  ✅       |
 | `@ViewChildren`                 |  ✅     |  ✅      |  ❌       |
-| `@ViewChild`                    |  ✅     |  ✅      |  ✅       |
+| `@ViewChild`                    |  ✅     |  ✅      |  ❌       |
```
Comment (mhevery) [inline packages/core/src/render3/STATUS.md]: Could we add the compiler changes? as follow on PR?

File: packages/core/src/render3/component.ts [hunk 2]
```diff
@@ -15,7 +15,7 @@ import {Sanitizer} from '../sanitization/security';
 import {assertComponentType, assertDefined} from './assert';
 import {queueInitHooks, queueLifecycleHooks} from './hooks';
 import {CLEAN_PROMISE, ROOT_DIRECTIVE_INDICES, _getComponentHostLElementNode, baseDirectiveCreate, createLViewData, createTView, detectChangesInternal, enterView, executeInitAndContentHooks, getRootView, hostElement, initChangeDetectorIfExisting, leaveView, locateHostElement, setHostBindings,} from './instructions';
-import {ComponentDef, ComponentDefInternal, ComponentType} from './interfaces/definition';
+import {ComponentDef, ComponentDefInternal, ComponentType, RenderFlags} from './interfaces/definition';
```
Comment (mhevery) [inline packages/core/src/render3/component.ts]: I think this Import is not needed

Comment (pkozlowski-opensource) [inline packages/core/src/render3/component.ts]: `RenderFlags` import left from other changes, remove.

File: packages/core/src/render3/definition.ts [hunk 3]
```diff
@@ -17,7 +16,7 @@ import {Type} from '../type';
 import {resolveRendererType2} from '../view/util';
 
 import {diPublic} from './di';
-import {ComponentDefFeature, ComponentDefInternal, ComponentTemplate, ComponentType, DirectiveDefFeature, DirectiveDefInternal, DirectiveDefListOrFactory, DirectiveType, DirectiveTypesOrFactory, PipeDef, PipeType, PipeTypesOrFactory} from './interfaces/definition';
+import {ComponentDefFeature, ComponentDefInternal, ComponentTemplate, ComponentType, DirectiveDefFeature, DirectiveDefInternal, DirectiveDefListOrFactory, DirectiveType, DirectiveTypesOrFactory, PipeDef, PipeType, PipeTypesOrFactory, RenderFlags} from './interfaces/definition';
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/definition.ts]: `RenderFlags` import not needed here, remove

Comment (mhevery) [inline packages/core/src/render3/definition.ts]: un-needed import revert

File: packages/core/src/render3/definition.ts [hunk 4]
```diff
@@ -126,6 +125,15 @@ export function defineComponent<T>(componentDefinition: {
...
   template: ComponentTemplate<T>;
 
+  /**
+   * Additional set of instructions specific to query processing. This could be seen as an
+   * "additional" template function.
+   * Query-related instructions need to be pulled out to a specific function as a timing of
+   * execution is different as compared to all other instructions (after change detection hooks but
+   * before view hooks).
+   */
+  queryInstructions?: ComponentTemplate<T>| null;
```
Comment (mhevery) [inline packages/core/src/render3/definition.ts]: bikeshedding: I think the word `instructions` does not add value. We don't call it `templateInstructions` and so it should be dropped.

Also even thought it is same signature as `ComponentTemplate` we should have a separate name alias for it. `ComponentQuery`

File: packages/core/test/render3/common_with_def.ts [hunk 5]
```diff
@@ -33,3 +33,13 @@ NgForOf.ngDirectiveDef = defineDirective({
   factory: () => new NgIfDef(injectViewContainerRef(), injectTemplateRef()),
   inputs: {ngIf: 'ngIf', ngIfThen: 'ngIfThen', ngIfElse: 'ngIfElse'}
 });
+
+(NgTemplateOutlet as any).ngDirectiveDef = defineDirective({
+  type: NgTemplateOutletDef,
+  selectors: [['', 'ngTemplateOutlet', '']],
+  factory: () => new NgTemplateOutletDef(injectViewContainerRef()),
+  features: [NgOnChangesFeature(
```
Comment (mhevery) [inline packages/core/test/render3/common_with_def.ts]: Could you rewrite `NgTemplateOutlet` to not have `NgOnChanges` in a follow up PR.

File: packages/core/test/render3/render_util.ts [hunk 6]
```diff
@@ -228,14 +228,16 @@ export function toHtml<T>(componentOrElement: T | RElement): string {
 
 export function createComponent(
     name: string, template: ComponentTemplate<any>, directives: DirectiveTypesOrFactory = [],
-    pipes: PipeTypesOrFactory = []): ComponentType<any> {
+    pipes: PipeTypesOrFactory = [],
```
Comment (mhevery) [inline packages/core/test/render3/render_util.ts]: number of positional args is getting long. Consider cleaning up by making the optional args named?

File: packages/core/test/bundling/hello_world/bundle.golden_symbols.json [hunk 7]
```diff
@@ -188,6 +188,12 @@
   {
     "name": "locateHostElement"
   },
+  {
+    "name": "queryCreateInstructions"
```
Comment (mhevery) [inline packages/core/test/bundling/hello_world/bundle.golden_symbols.json]: Hello world should not be pulling in these instructions. What is going on?

File: packages/core/test/bundling/todo/bundle.golden_symbols.json [hunk 8]
```diff
@@ -440,6 +440,12 @@
   {
     "name": "namespaceHTML"
   },
+  {
```
Comment (mhevery) [inline packages/core/test/bundling/todo/bundle.golden_symbols.json]: same here.

**Review by mhevery — APPROVED** [review]
I did most of the fixups here: 8413299f99
Can you take care of the remainder of the comments.

**Comment by pkozlowski-opensource** [comment]
@mhevery fixups from 8413299 LGTM - I was actually going back and forth on the `viewQuery` name :-)

If I understand correctly we can land this PR with your fixup and take care of the rest in the follow-up PRs, right? (I will add those items to the sprint tasks)

**Comment by mhevery** [comment]
> If I understand correctly we can land this PR with your fixup and take care of the rest in the follow-up PRs, right? (I will add those items to the sprint tasks)

correct, and please add them to sprint.

**Comment by mhevery** [comment]
http://test/OCL:201390514:BASE:201390558:1529520822838:454e5cbb

**Pull request closed without merging**